### PR TITLE
Bumping pandas version from 1.5.0 to 2.x

### PR DESCRIPTION
**Pull request by Yerzhaisang**
### Description
Pandas version 
 
### Issues Resolved
Closes https://github.com/opensearch-project/opensearch-py-ml/issues/270
 
### Check List
- [x] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by Yerzhaisang** [comment]
Dear @dhrubo-os, I upgraded pandas to the latest version and replaced all deprecated methods. I see opensearch dataframe uses datetime64[s], however it should use datetime64[ns]. Do you have any pieces of advice to suggest?

**Comment by dhrubo-os** [comment]
Add a changelog here: https://github.com/opensearch-project/opensearch-py-ml/blob/main/CHANGELOG.md


@rawwar could you please review this PR?

**Comment by Yerzhaisang** [comment]
> Add a changelog here: https://github.com/opensearch-project/opensearch-py-ml/blob/main/CHANGELOG.md
> 
> @rawwar could you please review this PR?

done

**Comment by rawwar** [comment]
@Yerzhaisang , I think the main idea is not to use a pin. Instead, use the version from the requirements.txt. Having this pin will lead to testing with a different version than the one user installs the package.  Can you try to remove the pins and check if tests are able to reuse the correct pandas version mentioned in the requirements.txt? Also, if tests are passing.

@dhrubo-os , are you fine with this?

**Comment by dhrubo-os** [comment]
> @Yerzhaisang , I think the main idea is not to use a pin. Instead, use the version from the requirements.txt. Having this pin will lead to testing with a different version than the one user installs the package. Can you try to remove the pins and check if tests are able to reuse the correct pandas version mentioned in the requirements.txt? Also, if tests are passing.
> 
> @dhrubo-os , are you fine with this?

Agree, I think that'll be the cleanest way. @rawwar could you please suggest @Yerzhaisang how to achieve this? Thanks.

File: .ci/run-tests [hunk 1]
```diff
@@ -5,7 +5,7 @@
 # Default environment variables
 export TEST_SUITE="${TEST_SUITE:=oss}"
 export PYTHON_VERSION="${PYTHON_VERSION:=3.9}"
-export PANDAS_VERSION=${PANDAS_VERSION-1.5.0}
+export PANDAS_VERSION=${PANDAS_VERSION-2.1.4}
```
Comment (rawwar) [inline .ci/run-tests:8]:
```suggestion

```

Comment (rawwar) [reply .ci/run-tests:8]: for starters, try to find out where this environment variable `PANDAS_VERSION` is being used and try to remove the usage.

File: noxfile.py [hunk 2]
```diff
@@ -103,7 +103,7 @@ def lint(session):
 
 
 @nox.session(python=["3.8", "3.9", "3.10"])
-@nox.parametrize("pandas_version", ["1.5.0"])
+@nox.parametrize("pandas_version", ["2.1.4"])
```
Comment (rawwar) [inline noxfile.py:106]:
```suggestion

```

**Comment by rawwar** [comment]
https://github.com/opensearch-project/opensearch-py-ml/blob/9b2f686cf2f3f5784173b5ec2f820c3adc8b892c/requirements.txt#L4 

https://github.com/opensearch-project/opensearch-py-ml/blob/9b2f686cf2f3f5784173b5ec2f820c3adc8b892c/requirements-dev.txt#L4

Do update the pandas version in above two places.

**Comment by Yerzhaisang** [comment]
Well, we see tests aren't able to reuse the correct pandas version mentioned in the requirements.txt, if we don't use the pin:
https://github.com/opensearch-project/opensearch-py-ml/actions/runs/7405262799/job/20147956089#step:3:873

Also, if we export latest version in tests and requirements.txt, we see assertion error:
https://github.com/opensearch-project/opensearch-py-ml/actions/runs/7405292325/job/20148022209#step:3:1155
Is it because the Opensearch dataframe uses datetime[s] type? Should it be datetime[ns]?

**Comment by rawwar** [comment]
> @rawwar thank you for the help!
> 
> We see tests aren't able to reuse the correct pandas version mentioned in the requirements.txt, if we don't use the pin: https://github.com/opensearch-project/opensearch-py-ml/actions/runs/7405262799/job/20147956089#step:3:873
> 

Along with removing the decorator, you need to remove pip install's for pandas.
currently, its `session.run("python", "-m", "pip", "install", f"pandas~={pandas_version}")`.  Can you delete this line and try?

**Comment by rawwar** [comment]
@Yerzhaisang , Tests are failing. Can you try removing pandas installations in the nox module?

**Comment by Yerzhaisang** [comment]
> @Yerzhaisang , Tests are failing. Can you try removing pandas installations in the nox module?

Hey, in ```noxfile.py``` 115th, 148th and 168th lines are already removed.

**Comment by Yerzhaisang** [comment]
Dear @rawwar and @dhrubo-os, thank you for the help. 

However, I don't understand, why are we trying to remove the pin? Initially, the issue wasn't about that. I mean https://github.com/opensearch-project/opensearch-py-ml/issues/270. We can fix the same pandas version in tests and requirements.txt and requirements-dev.txt, so we shouldn't see tests failures because of version diversity.

**Comment by rawwar** [comment]
> Dear @rawwar and @dhrubo-os, thank you for the help.
> 
> However, I don't understand, why are we trying to remove the pin? Initially, the issue wasn't about that. I mean #270. We can fix the same pandas version in tests and requirements.txt and requirements-dev.txt, so we shouldn't see tests failures because of version diversity.

I see the following error in the tests:

> 
> >>>>> Run [opensearch-project/opensearch-py-ml container] >>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> nox > Error while collecting sessions.
> nox > Sessions not found: test-3.9(pandas_version='')
> INFO: clean the network if not detached (start and exit)
> INFO: Removing container instance
> instance

I was asking to look into this.  What you said is correct. Hence, asking to check there was a miss in removing these pins in nox. However, I will also take a look into this today and respond

**Comment by rawwar** [comment]
@Yerzhaisang , give this a day or 2. I will take care of the removing the pin. You can continue on the primary issue once my PR is merged. I'm trying to remove pins and fix any issues with the current supported pandas version(`pandas>=1.5.2,<2`)

**Comment by Yerzhaisang** [comment]
> @Yerzhaisang , give this a day or 2. I will take care of the removing the pin. You can continue on the primary issue once my PR is merged. I'm trying to remove pins and fix any issues with the current supported pandas version(`pandas>=1.5.2,<2`)

Can you assign me your issue to me? I will work on that and if everything ok, I will raise another PR

**Comment by Yerzhaisang** [comment]
> > @Yerzhaisang , give this a day or 2. I will take care of the removing the pin. You can continue on the primary issue once my PR is merged. I'm trying to remove pins and fix any issues with the current supported pandas version(`pandas>=1.5.2,<2`)
> 
> Can you assign your issue to me? I will work on that and if everything ok, I will raise another PR.

I just suggest to collaborate in another draft PR

**Comment by rawwar** [comment]
> I just suggest to collaborate in another draft PR for your issue

Sure, here's the draft PR - https://github.com/opensearch-project/opensearch-py-ml/pull/368

**Comment by dhrubo-os** [comment]
@Yerzhaisang could you please expedite the development for this PR? Please let me know if you need any help? I'm planning to release a version for opensearch-py-ml after this PR merged into main.

**Pull request closed without merging**